cloudinit/signal_handler.py

```diff
@@ -5,28 +5,59 @@
 # Author: Joshua Harlow <harlowja@yahoo-inc.com>
 #
 # This file is part of cloud-init. See LICENSE file for license information.
+import contextlib
 import inspect
 import logging
 import signal
 import sys
 from io import StringIO
+from typing import Callable, Dict, Final, NamedTuple, Union
 
 from cloudinit import version as vr
 from cloudinit.log import log_util
 
 LOG = logging.getLogger(__name__)
 
-
-BACK_FRAME_TRACE_DEPTH = 3
-EXIT_FOR = {
-    signal.SIGINT: ("Cloud-init %(version)s received SIGINT, exiting...", 1),
-    signal.SIGTERM: ("Cloud-init %(version)s received SIGTERM, exiting...", 1),
-    # Can't be caught...
-    # signal.SIGKILL: ('Cloud-init killed, exiting...', 1),
-    signal.SIGABRT: ("Cloud-init %(version)s received SIGABRT, exiting...", 1),
+SIG_MESSAGE: Final = "Cloud-init %s received %s, exiting\n"
+BACK_FRAME_TRACE_DEPTH: Final = 3
+SIGNALS: Final[Dict[int, str]] = {
+    signal.SIGINT: "Cloud-init %(version)s received SIGINT, exiting",
+    signal.SIGTERM: "Cloud-init %(version)s received SIGTERM, exiting",
+    signal.SIGABRT: "Cloud-init %(version)s received SIGABRT, exiting",
 }
 
 
+class ExitBehavior(NamedTuple):
+    exit_code: int
+    log_level: int
+
+
+SIGNAL_EXIT_BEHAVIOR_CRASH: Final = ExitBehavior(1, logging.ERROR)
+SIGNAL_EXIT_BEHAVIOR_QUIET: Final = ExitBehavior(0, logging.INFO)
+SIGNAL_EXIT_BEHAVIOR = SIGNAL_EXIT_BEHAVIOR_CRASH
+
+
+def default_handler(_num, _stack) -> None:
+    """an empty handler"""
+    return None
+
+
+def inspect_handler(sig: Union[int, Callable, None]) -> None:
```

Comment:
Gather some data that would be useful for future debugging.

Reply:
Note that we won't actually see these logs unless we manually call cloud-init with `cloud-init --debug`. This is currently getting called pre-log setup.